pkg/dockerregistry/server/configuration/configuration.go

```diff
@@ -151,3 +252,210 @@ func pushEnv(environ []envVar) error {
+	return
+}
+
+func migrateCacheSection(cfg *Configuration, options configuration.Parameters) (err error) {
+	defBlobRepositoryTTL := defaultBlobRepositoryCacheTTL
+	defQuotaTTL := defaultProjectCacheTTL
+
+	if cfg.Cache != nil {
+		options = configuration.Parameters{}
+		defBlobRepositoryTTL = cfg.Cache.BlobRepositoryTTL
+		defQuotaTTL = cfg.Cache.QuotaTTL
+	} else {
+		cfg.Cache = &Cache{}
+	}
+
+	cfg.Cache.BlobRepositoryTTL, err = getDurationOption(BlobRepositoryCacheTTLEnvVar, "blobrepositorycachettl", defBlobRepositoryTTL, options)
+	if err != nil {
+		return
+	}
+	cfg.Cache.QuotaTTL, err = getDurationOption(ProjectCacheTTLEnvVar, "projectcachettl", defQuotaTTL, options)
+	return
+}
+
+func migratePullthroughSection(cfg *Configuration, options configuration.Parameters) (err error) {
+	defEnabled := true
+	defMirror := true
+
+	if cfg.Pullthrough != nil {
+		options = configuration.Parameters{}
+		defEnabled = cfg.Pullthrough.Enabled
+		defMirror = cfg.Pullthrough.Mirror
+	} else {
+		cfg.Pullthrough = &Pullthrough{}
+	}
+
+	cfg.Pullthrough.Enabled, err = getBoolOption(PullthroughEnvVar, "pullthrough", defEnabled, options)
+	if err != nil {
+		return
+	}
+	cfg.Pullthrough.Mirror, err = getBoolOption(MirrorPullthroughEnvVar, "mirrorpullthrough", defMirror, options)
+	return
+}
+
+func migrateCompatibilitySection(cfg *Configuration, options configuration.Parameters) (err error) {
+	defAcceptSchema2 := true
+
+	if cfg.Compatibility != nil {
+		options = configuration.Parameters{}
+		defAcceptSchema2 = cfg.Compatibility.AcceptSchema2
+	} else {
+		cfg.Compatibility = &Compatibility{}
+	}
+
+	cfg.Compatibility.AcceptSchema2, err = getBoolOption(AcceptSchema2EnvVar, "acceptschema2", defAcceptSchema2, options)
+	return
+}
+
+func migrateMiddleware(dockercfg *configuration.Configuration, cfg *Configuration) (err error) {
+	var repoMiddleware *configuration.Middleware
+	for _, middleware := range dockercfg.Middleware["repository"] {
+		if middleware.Name == middlewareName {
+			repoMiddleware = &middleware
+			break
+		}
+	}
+	if repoMiddleware == nil {
+		repoMiddleware = &configuration.Middleware{
+			Name:    middlewareName,
+			Options: make(configuration.Parameters),
+		}
+	}
+
+	if cfg.Auth == nil {
+		cfg.Auth = &Auth{}
+		cfg.Auth.Realm, err = getStringOption("", RealmKey, "origin", dockercfg.Auth.Parameters())
+		if err != nil {
+			return
+		}
+		cfg.Auth.TokenRealm, err = getStringOption("", TokenRealmKey, "", dockercfg.Auth.Parameters())
+		if err != nil {
+			return
+		}
+	}
+	if cfg.Audit == nil {
+		cfg.Audit = &Audit{}
+		authParameters := dockercfg.Auth.Parameters()
+		if audit, ok := authParameters["audit"]; ok {
+			auditOptions := make(map[string]interface{})
+
+			for k, v := range audit.(map[interface{}]interface{}) {
+				if s, ok := k.(string); ok {
+					auditOptions[s] = v
+				}
+			}
+
+			cfg.Audit.Enabled, err = getBoolOption("", "enabled", false, auditOptions)
```

Comment:
It would be nice to prefix such errors with the path to the parameter e.g.:

`configuration error in middleware->repository: failed to parse boolean option "enabled"`

My point is that there may be quite a lot of `enabled` options.

Reply:
@miminar Done

Reply:
Thanks. If `get<Type>Option` functions accepted full option path (e.g. `openshift.audit.enabled` instead of just `enabled`), the error messages could  be generated solely by `getOptionValue`.

But that's just a nit.